tests/src/metadata.rs

```diff
@@ -0,0 +1,257 @@
+        Ok(match s {
+            "Error" => AnnotationKind::Error,
+            "Warning" => AnnotationKind::Warning,
+            "Hint" => AnnotationKind::Hint,
+            "Autocomplete contains" => AnnotationKind::AutocompleteContains,
+            "Autocomplete excludes" => AnnotationKind::AutocompleteExcludes,
+            _ => return Err("invalid annotatino"),
+        })
+    }
+}
+
+impl Display for AnnotationKind {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.pad(self.as_str())
+    }
+}
+
+/// Parse metadata for a test.
+pub fn parse_part_metadata(source: &Source) -> TestMetadata {
+    let mut config = TestConfig::default();
+    let mut annotations = HashSet::default();
+    let mut invalid_data = vec![];
+
+    let lines: Vec<_> = source.text().lines().map(str::trim).collect();
+    let comments_until_code =
+        |i: usize| lines[i..].iter().take_while(|line| line.starts_with("//")).count();
+
+    let pos = |s: &mut Scanner, i: usize| -> Option<usize> {
+        let first = parse_num(s)? - 1;
+        let (delta, column) =
+            if s.eat_if(':') { (first, parse_num(s)? - 1) } else { (0, first) };
+        let line = (i + comments_until_code(i)).checked_add_signed(delta)?;
+        source.line_column_to_byte(line, usize::try_from(column).ok()?)
+    };
+
+    let range = |s: &mut Scanner, i: usize| -> Option<Range<usize>> {
+        s.eat_whitespace();
+        if s.eat_if("-1") {
+            let mut add = 1;
+            while let Some(line) = lines.get(i + add) {
+                if !line.starts_with("//") {
+                    break;
+                }
+                add += 1;
+            }
+            let next_line = lines.get(i + add)?;
+            let col = next_line.chars().count();
+
+            let index = source.line_column_to_byte(i + add, col)?;
+            s.eat_whitespace();
+            return Some(index..index);
+        }
+        let start = pos(s, i)?;
+        let end = if s.eat_if('-') { pos(s, i)? } else { start };
+        s.eat_whitespace();
+        Some(start..end)
+    };
+
+    for (i, line) in lines.iter().enumerate() {
+        let Some((key, value)) = parse_metadata_line(line) else { continue };
+        let key = key.trim();
+
+        match key {
+            "Ref" | "Hints" | "Autocomplete" => {
+                let value = value.trim();
+                if value != "false" && value != "true" {
+                    invalid_data.push((None, format!("Error: trying to set Ref, Hints, or Autocomplete with non-boolean: {value:?}.")));
+                }
+                match key {
+                    "Ref" => config.compare_ref = Some(value == "true"),
+                    "Hints" => config.validate_hints = Some(value == "true"),
+                    "Autocomplete" => {
+                        config.validate_autocomplete = Some(value == "true")
+                    }
+                    _ => unreachable!(),
+                }
+            }
```

Comment:
This duplication isn't nice and the `AnnotationKind::from_str(annotation_key).unwrap()` also not. I think the large match could match on each of the possible strings and dispatch to new subfunctions for parsing a boolean or a ranged message.

Reply:
I'd really love to use if let in guards to get rid of `AnnotationKind::from_str(annotation_key).unwrap()` unfortunately it's unstable :/

Reply:
Since we don't error on unknown keys anymore I wrote it in a nicer way `let Some(..) = .. else continue`